Add sample demonstrating VK_EXT_layer_settings usage

gpx1000 opened the pull request:
## Description

Demonstrate and provide a tutorial for using VK_EXT_layer_settings

Fixes #<issue number>

## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [ ] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [ ] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

gpx1000 committed:
Add sample demonstrating VK_EXT_layer_settings usage

SaschaWillems commented on samples/extensions/layer_settings/README.adoc:
```diff
@@ -0,0 +1,149 @@
+- limitations under the License.
+-
+////
+= VK_EXT_layer_settings — Configure validation layers programmatically
```
This is missing the "Where to get the source" Antora block (see readme.adoc for other samples or the readme from the samples template).

SaschaWillems commented:
This does work, but it feels a bit too minimalist. Wouldn't it make more sense to actually trigger some best practices thing (e.g. using wrong buffer flags or doing non-optimal transitions) and have them show up in the application's UI instead? Otherwise one has to scroll to the very top of the terminal to see some validation infos. And that does only work on platforms where you actually have a terminal.

gpx1000 committed:
- Update README with GitHub repository link for sample

gpx1000 committed:
Add local debug messenger and UI overlay for validation messages in LayerSettingsSample

gpx1000 commented:
I had thought keeping it simple was going to make it easier to take and use elsewhere; but went ahead and made the changes recommended.

gpx1000 committed:
Adjust code formatting in LayerSettingsSample for improved readability

gary-sweet commented:
So this is just supposed to be a black screen with some UI messages (that get cropped at the right side of the screen)?

asuessenbach commented on samples/extensions/layer_settings/README.adoc:
```diff
@@ -0,0 +1,153 @@
+- Enables the instance extension `VK_EXT_layer_settings` (optionally).
+- Adds settings for the Khronos validation layer before the Vulkan instance is created:
+  - Enables Best Practices checks.
+  - Enables `debugPrintfEXT` support (so shader `debugPrintf` messages are emitted via validation).
```
Those two bullet points are not rendered as sub-points.

asuessenbach commented on samples/extensions/layer_settings/README.adoc:
```diff
@@ -0,0 +1,153 @@
+        if (strcmp(e.extensionName, VK_EXT_LAYER_SETTINGS_EXTENSION_NAME) == 0)
+            return true;
+    }
+    return false;
```
Maybe:
`return std::ranges::find_if( exts, []( auto const & property ){ return strcmp( property.extensionName, VK_EXT_LAYER_SETTINGS_EXTENSION_NAME ) == 0; } ) != exts.end();`

asuessenbach commented on samples/extensions/layer_settings/README.adoc:
```diff
@@ -0,0 +1,153 @@
+    }
+    else
+    {
+        // Fallback for older SDKs: use VK_EXT_validation_features
```
No need to check for `VK_EXT_validation_features` support?

asuessenbach commented on samples/extensions/layer_settings/layer_settings.cpp:
```diff
@@ -0,0 +1,315 @@
+	{
+		VkInstance instance  = get_instance().get_handle();
+		auto       fpDestroy = reinterpret_cast<PFN_vkDestroyDebugUtilsMessengerEXT>(
+            vkGetInstanceProcAddr(instance, "vkDestroyDebugUtilsMessengerEXT"));
```
Why do you explicitly get the function pointer? As volk is initialized by the framework, just calling
`		vkDestroyDebugUtilsMessengerEXT(instance, debug_messenger_, nullptr);
`works as well

asuessenbach commented on samples/extensions/layer_settings/layer_settings.cpp:
```diff
@@ -0,0 +1,315 @@
+
+	// Install a local DebugUtils messenger to collect validation messages into the UI.
+	// If it was already created during create_instance(), skip to avoid duplication.
+	if (debug_messenger_ == VK_NULL_HANDLE && get_instance().is_enabled(VK_EXT_DEBUG_UTILS_EXTENSION_NAME))
```
I don't get this.
Under what circumstances would the `debug_messenger_` not be created during `create_instance`, which is called before this function?

asuessenbach commented on samples/extensions/layer_settings/layer_settings.cpp:
On my configuration (Win10, NVIDIA GPU), this sample never gets to appending some line to the log_text_.

gpx1000 committed:
Merge branch 'main' into VK_EXT_layer_settings

# Conflicts:
#	samples/extensions/README.adoc

gpx1000 committed:
Add interactive validation scenarios with toggleable UI and message caching

Transforms layer_settings sample into an interactive demonstration with three toggleable validation scenarios (wrong buffer flags, suboptimal transitions, small allocations). Executes validation-triggering operations once during setup using one-time command buffers to work with message caching. Adds per-scenario warning/error counters, message caching, and resource cleanup. Updates UI with checkboxes, statistics display, and scrollable message viewer with improved styling.

gpx1000 committed:
Update copyright year to 2026

asuessenbach commented on samples/extensions/layer_settings/layer_settings.cpp:
```diff
@@ -278,35 +589,241 @@ VKAPI_ATTR VkBool32 VKAPI_CALL LayerSettingsSample::debug_callback(
-	snprintf(line, sizeof(line), "[%s][%s] %s (%s)\n", sev, type_buf[0] ? type_buf : "-",
-	         pCallbackData && pCallbackData->pMessage ? pCallbackData->pMessage : "<no message>",
-	         pCallbackData && pCallbackData->pMessageIdName ? pCallbackData->pMessageIdName : "no-id");
+	snprintf(line, sizeof(line), "[%s][%s] %s\n", sev, type_buf[0] ? type_buf : "-",
```
Maybe use `std::format`?

asuessenbach commented on samples/extensions/layer_settings/layer_settings.h:
```diff
@@ -38,18 +40,70 @@ class LayerSettingsSample : public ApiVulkanSample
 
-	// Local debug messenger to capture Validation Layer messages at runtime.
+	// Per-scenario state tracking
+	std::array<ScenarioState, static_cast<size_t>(Scenario::Count)> scenario_states_;
```
Maybe using a `std::unordered_map<Scenario, ScenarioState>` would be a better choice here: you don't need that artificial `Scenario::Count`, and you don't need those `static_cast`s to access the corresponding `ScenarioState`.

gpx1000 committed:
Refactor scenario state tracking from array to unordered_map and modernize C++ usage

Replaces `std::array` with `std::unordered_map<Scenario, ScenarioState>` for scenario state storage, removing the `Scenario::Count` sentinel. Adds custom hash functor for the enum key. Modernizes code with `std::format`, `std::ranges`, and direct use of `vkCreateDebugUtilsMessengerEXT`. Updates README list formatting, adds `validation_features_supported()` helper, and initializes scenario map entries in constructor. Improves code consistency with better whitespace and comment cleanup.

gpx1000 commented:
To be clear, when you toggle the three items in the UI, you will see the warnings assiated with those 3 items.  The background is clear as I thought this the best way to demonstrate the feature.

gpx1000 committed:
Fix indentation to use tabs consistently throughout layer_settings files

gpx1000 commented:
Successful CI Building will depend upon the updated clang tidy call in #1418 Please ensure that is taken before validating that this one is CI clean.

gary-sweet commented:
This has the same std::format build problems for me as the pipeline_binary sample. We'll need to discuss (and hopefully undo) those format changes so that Broadcom can build the samples.

asuessenbach commented on samples/extensions/layer_settings/layer_settings.cpp:
```diff
@@ -774,15 +745,16 @@ void LayerSettingsSample::on_update_ui_overlay(vkb::Drawer &drawer)
 		// Check if all scenarios are disabled and clear log if so
 		bool any_enabled = false;
-		for (const auto &state : scenario_states_)
+		for (const auto &kv : scenario_states_)
```
What does `kv` stand for?

gpx1000 replied:
Key / Value  I'll change it to something more appropiate.

gpx1000 committed:
Replace std::format with fmt::format and refactor loop variable naming

Removes unused `<format>` include and switches to `fmt::format` for formatting. Renames loop variable from `kv` to `scenario` for better clarity when iterating over scenario states map.

gary-sweet commented:
This is another one that only actually works properly when build in debug mode.
We should make that obvious somewhere to avoid confusion.

gpx1000 committed:
Merge branch 'main' into VK_EXT_layer_settings

gpx1000 committed:
Add note requested by comments.

gpx1000 committed:
Add instance extension and layer setting registration to VulkanSample base class

Adds `add_instance_extension()` and `add_layer_setting()` methods to allow samples to register instance-level extensions and layer settings. Introduces `instance_extensions` map and `layer_settings` vector as member variables. Integrates these into `request_instance_extensions()` and `request_layer_settings()` to automatically apply sample-specific configurations during instance creation.

gpx1000 committed:
Fix whitespace alignment in vulkan_sample.h and correct method name in layer_settings

Aligns `present_mode` and `present_mode_priority_list` declarations with consistent spacing. Replaces incorrect `is_enabled()` call with `is_extension_enabled()` for debug utils extension check.

asuessenbach commented on framework/vulkan_sample.h:
This file has been completely changed? Or not at all?

gpx1000 replied:
CRLF to LF.  line endings change.  I'm trying to get git to allow me to go back to CRLF :(

gpx1000 replied:
Okay, now it should correctly show changes only.

gpx1000 committed:
Merge branch 'refs/heads/main' into VK_EXT_layer_settings

# Conflicts:
#	antora/modules/ROOT/nav.adoc
#	framework/vulkan_sample.h
#	samples/extensions/README.adoc

gpx1000 committed:
Revert the CRLF to Unix-style line endings that caused the whole file change inadvertently.

gpx1000 committed:
Merge branch 'main' into VK_EXT_layer_settings

tomek-brcm commented:
The updated change works for Broadcom. you might wait for one more review though.

asuessenbach commented on framework/vulkan_sample.h:
```diff
@@ -246,6 +246,10 @@ class VulkanSample : public vkb::Application
 	 */
 	void add_device_extension(const char *extension, bool optional = false);
 
+	void add_instance_extension(const char *extension, bool optional = false);
```
The current approach to add instance extensions is to overload `VulkanSample::request_instance_extensions`.

asuessenbach replied:
Not that this function isn't used any more, you should not add it.

asuessenbach commented on framework/vulkan_sample.h:
```diff
@@ -246,6 +246,10 @@ class VulkanSample : public vkb::Application
 
+	void add_instance_extension(const char *extension, bool optional = false);
+
+	void add_layer_setting(const LayerSettingType &layer_setting);
```
And the current approach to add layer settings is to overload `VulkanSample::request_layer_settings`.

asuessenbach replied:
As above.

asuessenbach commented on framework/vulkan_sample.h:
```diff
@@ -410,6 +414,12 @@ class VulkanSample : public vkb::Application
 	std::unordered_map<const char *, bool> device_extensions;
 
+	/** @brief Set of instance extensions to be enabled for this example and whether they are optional (must be set in the derived constructor) */
+	std::unordered_map<const char *, bool> instance_extensions;
```
Those `instance_extensions` should also not be added.

asuessenbach commented on framework/vulkan_sample.h:
```diff
@@ -410,6 +414,12 @@ class VulkanSample : public vkb::Application
+	std::unordered_map<const char *, bool> instance_extensions;
+
+	/** @brief Set of layer settings to be enabled for this example (must be set in the derived constructor) */
+	std::vector<LayerSettingType> layer_settings;
```
Those `layer_settings` should not be added.

asuessenbach commented on samples/extensions/layer_settings/README.adoc:
```diff
@@ -0,0 +1,191 @@
+Demonstrates the importance of declaring correct usage flags at buffer creation time. Using a buffer without the appropriate usage bit (e.g., `VERTEX_BUFFER_BIT`) can lead to undefined behavior or performance issues on some implementations.
+
+*Suboptimal Transitions*::
+Shows why choosing the right image layout matters. Transitioning to `GENERAL` layout is legal but often suboptimal—specific layouts like `COLOR_ATTACHMENT_OPTIMAL` or `SHADER_READ_ONLY_OPTIMAL` enable better performance.
```
I don't know on how many platforms `GENERAL` actually is suboptimal... so maybe just "... is legal but potentially suboptimal ..."?

asuessenbach commented on samples/extensions/layer_settings/README.adoc:
```diff
@@ -0,0 +1,191 @@
+#include <cstring>
+#include <ranges>
+
+static const char *kValidationLayer = "VK_LAYER_KHRONOS_validation";
```
Maybe `kValidationLayer` -> `kValidationLayerName`?
